src/concurrent_server.py

```diff
@@ -0,0 +1,162 @@
+from __future__ import unicode_literals
+
+import socket
+import email.utils
+import mimetypes
+import io
+import os
+
+
+def build_file_structre_html(directory):
+    print(directory)
+    body = "<!DOCTYPE html>\r\n<html>\r\n<ul>\r\n<body>\r\n<h1>Code Fellows</h1>\r\n"
+    for item in os.listdir(directory):
+        print(item)
+        if os.path.isdir(os.path.join(directory, item)):
+            body += "<li><a href=\"{}/\">{}</a></li>\r\n".format(item, item)
+        else:
+            body += "<li><a href=\"{}\">{}</a></li>\r\n".format(item, item)
+    body += "</ul>\r\n</body>\r\n</html>"
+    return body
+
+
+def response_template():
+    return [u"",
+            u"" + str("Date: " + email.utils.formatdate(usegmt=True) + "\r\n"),
+            u"Content-type: text/html; charset=utf-8\r\n",
+            u"Content-length: \r\n\r\n",
+            u"Body: "]
```

Comment:
What if instead of returning a list of strings that all end with `\r\n` you tried just returning a list of strings:

```python
def response_template():
    lines = [
        b"",
        b"Date: " + email.utils.formatdate(usegmt=True).encode('utf8'),
        b"Content-type: text/html; charset=utf-8",
        b"Content-length: ",
        b"",
    ]
    return lines
```

Then you could do things like append the body to this list as a new "line", insert the correct content type line in place of the one already there, append the length of the body to the content length line, and even insert additional headers beyond those already in your template.

Finally, when you were ready to send it, you could take care of making sure that all the lines were properly delimited, you could do something like this:

```python
conn.sendall(b"\r\n".join(lines))
```